text/0000-custom-model-classes.md

```diff
@@ -0,0 +1,158 @@
+- Start Date: 2019-05-09
+- Relevant Team(s): data 
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
+- Tracking: (leave this empty)
+
+# Custom Model Class RFC
+
+## Summary
+
+This RFC is a follow-up RFC for #293 RecordData and replaces [https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor](https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor)
+
+Create a pluggable method for instantiating model classes. Add an `instantiateRecord` method to `Store` which would allow addons and ED itself to offer a clean replacement the default DS.Model with a custom Model class. 
+
+## Motivation
+
+- Allowing Addons to experiment with replacing the default DS.Model implementation ([https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor](https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor))
+- First step in separating the record implementation from schema information
+
+## Detailed design
+
+After the work in [https://www.notion.so/Errors-RFC-8e9e1fd4e2474ed5a8539ed343047784](https://www.notion.so/Errors-RFC-8e9e1fd4e2474ed5a8539ed343047784),  [https://www.notion.so/Request-State-RFC-b994c09fe3094f2b92e4ff1dd36fa882](https://www.notion.so/Request-State-RFC-b994c09fe3094f2b92e4ff1dd36fa882), [https://www.notion.so/Record-State-on-Record-Data-RFC-3fd05bcd2a83439f96fbcf2bc0f3a581](https://www.notion.so/Record-State-on-Record-Data-RFC-3fd05bcd2a83439f96fbcf2bc0f3a581) we have enough coverage in public apis that we can implement the entire DS.Model class with just a few changes to Store APIs. 
+
+We will need four separate changes: 
+
+- A way to instantiate custom Records and have those records be notified of changes in Record Data
+- A way to access schema information not from the Record Class
+- Replacement of few store methods that take DS.Models
+- Deprecate passing DS.Model classes to adapter/serializers/exposing them on Snapshots
```

Comment:
Presumably we will depend on a feature flag for this?  We should call it out to make it clearer to custom model implementers what they need to do

Reply:
Per the discussion at the weekly meeting, this is implied.